File: lib/notification.ex
defmodule Notification do
  require Logger
  import Cfg

  @doc """
  Sends notification using native OSX Notification Center.
  For Tmux sessions, Small includes helper application to fix issues with Notification Center not working under Tmux. More: https://github.com/ChrisJohnsen/tmux-MacOSX-pasteboard.

  ## Examples

      iex> Notification.send "abc"
      :ok

  """
  @spec send(String.t) :: :ok | :error
  @spec send(String.t, sound_name :: Atom.t()) :: :ok | :error
  def send(message, sound_name \\ :no_sound) do
    sound_command = if sound_name == :no_sound, do: "", else: "sound name \"#{sound_name}\""

    case System.cmd("/usr/bin/osascript", [
           "-e",
           "display notification \"#{message}\" #{sound_command} with title \"Small\""
         ]) do
      {_, 0} ->
        :ok

      {_, _} ->
        :error
    end
  end

  @doc """
  Send notifications based on user settings.
  Notify with error level will automatically raise an exception.
  Types: [:start, :clipboard, :upload, :error]
  """
  @spec notification(String.t, Term.t()) :: nil
  def notification(message, type) do
    config = config()

    if config[:notifications][type] do
      Logger.debug(
        "Notification of type #{inspect(type)} with result: #{inspect(config[:notifications][type])}"
      )

      if config[:sounds][type] do
        Notification.send(
          message,
          config[:sounds][String.to_atom(Atom.to_string(type) <> "_sound")]
        )
      else
        Notification.send(message)
      end

      if type == :error do
        raise message
      end
    end
  end
end

File: src/notification.rs
use crate::{
    config::{NotificationSettings, SoundSettings},
    *,
};
use anyhow::Result;
use std::process::Command;


pub fn send(message: &str, sound_name: Option<&str>) -> Result<()> {
    let sound_command = if let Some(sound) = sound_name {
        format!("sound name \"{sound}\"")
    } else {
        String::new()
    };

    let script =
        format!("display notification \"{message}\" {sound_command} with title \"Small\"");
    let output = Command::new("/usr/bin/osascript")
        .args(["-e", &script])
        .output()?;

    if output.status.success() {
        Ok(())
    } else {
        anyhow::bail!("Failed to send notification")
    }
}


pub fn notification(
    message: &str,
    notification_type: &str,
    notifications: &NotificationSettings,
    sounds: &SoundSettings,
) -> Result<()> {
    let should_notify = match notification_type {
        "start" => notifications.start,
        "clipboard" => notifications.clipboard,
        "upload" => notifications.upload,
        "error" => notifications.error,
        _ => false,
    };

    if !should_notify {
        return Ok(());
    }

    debug!("Notification of type {notification_type} with result: {should_notify}");
    let sound_name = match notification_type {
        "start" if sounds.start => Some(sounds.start_sound.as_str()),
        "clipboard" if sounds.clipboard => Some(sounds.clipboard_sound.as_str()),
        "upload" if sounds.upload => Some(sounds.upload_sound.as_str()),
        "error" if sounds.error => Some(sounds.error_sound.as_str()),
        _ => None,
    };

    send(message, sound_name)?;
    if notification_type == "error" {
        anyhow::bail!("{message}");
    }

    Ok(())
}
